keylime/cert_utils.py

```diff
@@ -71,46 +76,78 @@ def verify_ek(ekcert: bytes, tpm_cert_store: str) -> bool:
                     )
                 elif isinstance(signcert_pubkey, EllipticCurvePublicKey):
-                    assert ek509.signature_hash_algorithm is not None
+                    assert cert.signature_hash_algorithm is not None
                     signcert_pubkey.verify(
-                        ek509.signature,
-                        ek509.tbs_certificate_bytes,
-                        ec.ECDSA(ek509.signature_hash_algorithm),
+                        cert.signature,
+                        cert.tbs_certificate_bytes,
+                        ec.ECDSA(cert.signature_hash_algorithm),
                     )
                 else:
                     logger.warning("Unsupported public key type: %s", type(signcert_pubkey))
                     continue
             except crypto_exceptions.InvalidSignature:
                 continue
 
-            logger.debug("EK cert matched cert: %s", cert_file)
+            logger.debug("Cert to verify matched cert: %s", cert_file)
             return True
     except Exception as e:
         # Log the exception so we don't lose the raw message
         logger.exception(e)
-        raise Exception("Error processing ek/ekcert. Does this TPM have a valid EK?").with_traceback(sys.exc_info()[2])
+        raise Exception("Error processing %s certificate.", cert_type).with_traceback(sys.exc_info()[2])
 
-    logger.error("No Root CA matched EK Certificate")
+    logger.error("No Root CA matched %s Certificate", cert_type)
     return False
 
 
+def check_tpm_origin(cert: Certificate, cert_type: str = "") -> bool:
+    """Verify that the provided certificate is from a TPM according to the SAN
+    :param cert: The certificate as a cryptography.x509.Certificate
+    :param cert_type: Type of certificate as string for logging
+    :returns: True if the certificate came from a TPM
+    """
+    san_ext = cert.extensions.get_extension_for_oid(ExtensionOID.SUBJECT_ALTERNATIVE_NAME)
+    othername = san_ext.value.get_values_for_type(x509.OtherName)[0]
+    if not othername.type_id.dotted_string == OID_HW_MODULE_NAME:
+        logger.error("%s Certificate did not contain the HW module name. This may not have come from a TPM.", cert_type)
+        return False
+    decoded_on, _ = decoder.decode(othername.value, asn1Spec=rfc4108.HardwareModuleName())
+    if not str(decoded_on['hwType']) == OID_HWTYPE_TPM:
+        logger.error("%s Certificate did not contain the correct hwType OID in SubjectAltName. This may not have come from a TPM.", cert_type)
+        return False
+    return True
+
+
+def verify_ek(ekcert: bytes, tpm_cert_store: str) -> bool:
+    """Verify that the provided EK certificate is signed by a trusted root
+    :param ekcert: The Endorsement Key certificate in DER format
+    :param tpm_cert_store: The path for the TPM certificate store
+    :returns: True if the certificate can be verified, False otherwise
+    """
+    try:
+        ek509 = x509_der_cert(ekcert)
+    except Exception as e:
```

Comment:
(seems that this `e` will raise a complain, so `err`)

Reply:
should it be this:
```
except Exception as e:
    logger.error("Error processing the EK certificate: %s", e)
    return False
```

or do you mean something else?

Reply:
He mean to replace `e` with `err` since some linters (that are not active yet) will complain about variables that are too short.